src/base/kaldi-error.h

```diff
@@ -78,81 +80,100 @@ struct LogMessageEnvelope {
   int32 line;
 };
 
-// Class MessageLogger is invoked from the KALDI_ASSERT, KALDI_ERR, KALDI_WARN and
-// KALDI_LOG macros. It formats the message, then either prints it to stderr or
-// passes to the log custom handler if provided, then, in case of the error,
-// throws an std::runtime_exception, in case of failed KALDI_ASSERT calls abort().
+/// Kaldi runtime error exception. Thrown from any use of KALDI_ERR.
+class KaldiFatalError : public std::runtime_error {
+ public:
+  explicit KaldiFatalError(const std::string &message)
+      : std::runtime_error(message) { }
+  explicit KaldiFatalError(const char *message)
+      : std::runtime_error(message) { }
+
+  //TODO(kkm): Temporary(?) hack. Do we really need it? I think a better
```

Comment:
if this is going to be permanent, better to just explain the purpose,

Reply:
Sure thing, I want to remove the TODO and finalize the interface before this merge. We need to decide how this will be used. My complete question is under the "NEED FEEDBACK PLEASE" heading in the description. Very briefly, I do not like the current approach, which is: log the error then throw runtime_error(""). My understanding this empty string is to avoid double-printing of the already logged message, because all bins do `cerr << e.what();`. The two member functions of the KaldiFatalError exist solely to continue to support this hack: `what()` always returns "", but the additional member KaldiMessage() is available (just in case) to get the error message.

What I would like to do is change the catch clause in all binaries to `catch (const KaldiFatalError&) { return -1; }`, essentially. In this case, the `what()` hack becomes unnecessary, and the KaldiMessage() accessor won't be needed (so no non-inherited members in the class KaldiFatalError). It's a non-zero amount of work, naturally, and I wanted to run it by you first. And what should be the form of the catch-clause wrt catching non-Kaldi std:: exceptions, see the full question for the two possibilities.

Reply:
Or maybe I am not understanding the reason for the empty what() message at all?

Reply:
I make use of catching exceptions with a custom logger and thus don't assume the message has gone to the output.  Does this effect your thoughts on your "NEED FEEDBACK PLEASE" design ?

Reply:
I also do, although I know I should not; Kaldi has never been designed for that. At some points it leaks memory, for one. 

I am not changing any existing behavior. The custom logger (or the default one, no matter) is called before throwing, and has always been. And if it's called with kError, you know it's going to throw. It it's called with kAssertFailed, Kaldi will abort, but you can throw your own exception at this point and bypass the abort. Whether to log or not, your choice in the logging function. Of course, having caught the exception, you should assume that all the state has been corrupted, memory leaked and gremlins are out to get you.